### DISCUSSION TITLE
Datepicker component

### DISCUSSION THREAD

**Discussion by aaronBery** [Legal-and-General/canopy#207] [General] [answered]
The ideal solution would have been to harness the native date picker however this isn't supported in all major browsers including Safari on MacOS.

One solution would be to create our own custom date picker. We would have full control over the component and we could keep the look and feel inline with other Canopy components.

The main consideration with this approach will however be how we handle mobile devices. Mobile users receive an excellent experience with native date pickers so any solution we create would need to offer a mobile friendly experience.

Alternatively we could detect mobile devices and fallback to the native picker. However there would be an overhead in detecting which device receives the native experience and could over complicate the component and possibly introduce external dependencies in the form of device detection libraries.

If we were to start developing a custom date picker component we would need an agreed design to capture the look and feel of the component (which takes into account our approach for handling mobile devices). 

Some technical considerations are that we currently use `date-fns` so it would make sense to use this in scenarios where we need to use date based functions.

**Comment by elenagarrone** [comment]
@aaronBery I would move this to the `Discussions` tab

**Comment by andy-polhill** [comment]
I agree with the approach of using the native date picker on mobile devices and then a custom one on desktop devices. I'm not sure of the most effective way to determine that though. I think there is some work to do there to find the most efficient way.

So far in Canopy we don't have any floating elements, personally I'm quite happy with that, but the date picker might be a candidate for that. If that is the case it could be worth not having them floating in the first iteration, for simplicity sake.

**Reply by aaronBery** [reply]
It might just be worth bearing in mind other libraries such as Angular Material don't fall back to a native picker on mobile https://material.angular.io/components/datepicker/overview 

although I don't think they do a great job in terms of usability on mobile....if we went down a similar path we could put more effort into optimising for smaller devices.

In terms of the device detection (with the fallback to native) approach there are a few articles solutions out there that interrogate the `navigator.userAgent` and look to see if it contains things like 'Android' or 'iPhone' but there are suggestions this approach is unreliable (see https://developer.mozilla.org/en-US/docs/Web/API/NavigatorID/userAgent). So as you say the biggest barrier to this approach is finding a reliable approach for detecting mobile devices.

**Reply by paul-request** [reply]
I always thought if we needed device detection, then we would possibly use a third party service that would maintain that list with new devices etc getting added. Personally, I think if we don't detect the device, then we would really only be able to offer native experience to small devices, as the lines between desktop and tablet are way too blurred

**Reply by paul-request** [reply]
As a result, the simplest thing to do would be not to offer a native experience at all IMO. Which would be a shame... I wonder what other use cases there could be for device detection in the future of Canopy 🤔

**Reply by aaronBery** [reply]
> I always thought if we needed device detection, then we would possibly use a third party service that would maintain that list with new devices etc getting added. Personally, I think if we don't detect the device, then we would really only be able to offer native experience to small devices, as the lines between desktop and tablet are way too blurred

I guess the problem arises when a Safari user is using a browser window that falls under the category of small device and we attempt to show a native picker that doesn't exist. Or an iOS or Android tablet user has a larger screen that crosses a certain breakpoint and they loose the native picker.

**Reply by aaronBery** [reply]
I should have added this is a problem if we use screen size to determine which experience we provide i.e. native vs our custom implantation

**Comment by owensgit** [comment]
So far it seems that:

- We will need to build a custom component for at least tablet and desktop devices
- Using native picker on small devices may provide troublesome and unreliable

I wonder if we should attempt to build a custom component that will work on all devices, potentially improving on the usability of Material's (tough gig - but I think we could have a good go at it!)

Perhaps we could also research ways of showing native date-pickers on mobile devices - if we can find a reliable and lightweight way of doing it then it will be a great enhancement - and if not, at least we will have a custom component that's the best it can be on mobile.

Of course we'll also need to consider how assistive technology will work too.

Whichever way, this is a good challenge for us!

**Reply by owensgit** [reply]
If we have a decent custom component as a fallback, and we're able to reliably target 80%+ of mobile devices to show native picker, I think we'd be doing quite well.

**Reply by paul-request** [reply]
I suppose as an MVP, a custom Datepicker for all devices sounds like the easiest and simplest thing to get started with. Can always continue the discussion around how to support natively if we feel there is a need. If we have more use cases for it, perhaps it would be worth investing the time

**Reply by aaronBery** [reply]
Agreed @owensgit we can probably build a custom date picker with a better user experience for users on mobile devices :-) 

I also agree with @paul-request  in terms of building a custom picker for all devices. I have reservations about device detection as it's unreliable and as Paul says the lines are blurred and not to mention ever evolving. 

The complete opposite approach would be to go fully native and with a fall back to a text input for browsers that don't support the native date (which once IE11 dies is just Safari right?)

**Reply by owensgit** [reply]
> Can always continue the discussion around how to support natively if we feel there is a need. 

Absolutely, I think it something that could be looked at in parallel and added as an enhancement if a good solution can be found 👍

**Reply by aaronBery** [reply]
I should add the fallback to text for a fully native approach might be nasty as we would have additional work to parse the text into a date object whereas in our fully custom approach we can maintain the same object type

**Reply by andy-polhill** [reply]
> I suppose as an MVP, a custom Datepicker for all devices sounds like the easiest and simplest thing to get started with

I agree with this, I think a custom date picker helps if we want to do additional things like date range selection as well.

**Reply by elenagarrone** [reply]
I agree with building a custom date picker for all the screen sizes; I wouldn't go down the route of native date pickers at a different breakpoint - it will bring more bugs than anything else.
Personally I think the Material one works perfectly on all devices and is approved by [DIG](https://diginclusion.com/) too accessibility wise so I would follow that approach.

Keep in mind that for accessibility the preferred way of entering a date is via an input text. Very rarely screen reader users will go inside the date picker modal so we should offer both. This is what we currently have in our project:
![Screenshot 2021-02-09 at 09 06 17](https://user-images.githubusercontent.com/8397116/107341223-9e751f00-6ab6-11eb-99f2-54f0426499f4.png)

**Reply by aaronBery** [reply]
Thanks for the feedback @elenagarrone, really helpful especially surrounding the accessibility. Maybe I have stubby fingers but I find you have to be very precise with the Angular Material date picker on mobile so this was one area as a user I felt we could improve (if we are following a similar design to theirs)...works really well on desktop though :-)

**Reply by owensgit** [reply]
Thanks @elenagarrone that's great to know the Material one is approved by Digg 😃 

> I wouldn't go down the route of native date pickers at a different breakpoint - it will bring more bugs than anything else.

Just to clarify the idea round this; I would imagine the preferred approach would be to try and find a lightweight and reliable solution for device detection, possibly via the user agent string. And if it one can be found, enable the native solution on mobile devices as a kind of progressive enhancement. This could be done as an investigation in parallel - but of course may not turn out to be realistic - but I think it's worth a shot.

**Reply by alexcanning** [reply]
I'm gonna be contrary and ask why we should go through the effort of designing and developing a custom solution when there are decent native solutions on desktop and mobile? On mobile, it's usually best to leverage the native component if you can - it's the thing that users are gonna be most familiar with. It'll have all the native interactions and behaviours built in, and the user will have encountered it many times elsewhere across their mobile experience. I'm concerned there's a lot of accessibility, logic and interaction behaviours we need to consider.

Safari being a sticking point (if we discount IE11) - the current desktop version doesn't have a picker, but at least it degrades to a basic input. The iOS version has a pretty decent date picker. I can also see from Safari Technology Preview (desktop) that there is a date picker on the way. 

What's our share of users for Safari on desktop? If it's low enough, and the fix is coming anyway, is there value in us developing and building something that would be superseded by a native solution in the near future?

Apologies if I've re-opened old wounds or misunderstood the point being made!

**Reply by michelleallendavey** [reply]
Hello all – I'm picking up the date picker design, but just noticed @alexcanning's comment above. Has there been any more conversation or progress made since then? (Doesn't matter at all if not – just want to make sure I won't be doing unnecessary work) @owensgit @aaronBery @elenagarrone

**Reply by owensgit** [reply]
@alexcanning @michelleallendavey I think in any case we're going to need a custom date picker in order to support browsers which don't have one. One issue we have is that there's not a water-tight method of detecting if a browser has a native picker, so one suggestion is to create the custom picker anyway and then put some time into researching how to provide the native solution if it exists.

**Comment by owensgit** [comment, marked as answer]
From reading all the comments, I think the consensus is that:

1. We build a custom date picker
2. Separately, we research ways of detecting if a native date picker is present and use that instead, if a reliable solution can be found.

Please up-vote if you agree and I'll mark this as the answer 💥
